controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -855,3 +867,131 @@ func (r *KubeadmControlPlaneReconciler) ClusterToKubeadmControlPlane(o handler.M
+			r.recorder.Eventf(kcp, corev1.EventTypeWarning, "AdoptionFailed", "Could not adopt Machine %s/%s: its version (%q) is outside supported +/- one minor version skew from KCP's (%q)", m.Namespace, m.Name, *m.Spec.Version, kcp.Spec.Version)
+			// avoid returning an error here so we don't cause the KCP controller to spin until the operator clarifies their intent
+			return nil
+		}
+	}
+
+	for _, m := range machines {
+		ref := m.Spec.Bootstrap.ConfigRef
+		cfg := &bootstrapv1.KubeadmConfig{}
+		err := r.Client.Get(ctx, client.ObjectKey{Name: ref.Name, Namespace: kcp.Namespace}, cfg)
+		if err != nil {
+			return err
+		}
+
+		err = r.adoptOwnedSecrets(ctx, kcp, cfg)
+		if err != nil {
+			return err
+		}
+
+		patchHelper, err := patch.NewHelper(m, r.Client)
+		if err != nil {
+			return err
+		}
+
+		if err = controllerutil.SetControllerReference(kcp, m, r.scheme); err != nil {
+			return err
+		}
+
+		// 0. get machine.Spec.Version - the easy answer
+		machineKubernetesVersion := ""
+		if m.Spec.Version != nil {
+			machineKubernetesVersion = *m.Spec.Version
+		}
+
+		// 1. hash the version (kubernetes version) and kubeadm_controlplane's Spec.infrastructureTemplate
+		syntheticSpec := controlplanev1.KubeadmControlPlaneSpec{
+			Version:                machineKubernetesVersion,
+			InfrastructureTemplate: kcp.Spec.InfrastructureTemplate,
+			KubeadmConfigSpec:      equality.SemanticMerge(cfg.Spec, kcp.Spec.KubeadmConfigSpec, cluster),
+		}
+		newConfigurationHash := hash.Compute(&syntheticSpec)
+		// 2. add kubeadm.controlplane.cluster.x-k8s.io/hash as a label in each machine
+		m.Labels["kubeadm.controlplane.cluster.x-k8s.io/hash"] = newConfigurationHash
+
+		// Note that ValidateOwnerReferences() will reject this patch if another
+		// OwnerReference exists with controller=true.
+		if err := patchHelper.Patch(ctx, m); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *KubeadmControlPlaneReconciler) adoptOwnedSecrets(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, currentOwner metav1.Object) error {
+	secrets := corev1.SecretList{}
+	if err := r.Client.List(ctx, &secrets, client.InNamespace(kcp.Namespace)); err != nil {
+		return errors.Wrap(err, "error finding secrets for adoption")
+	}
+
+	for _, s := range secrets.Items {
+		if !util.PointsTo(s.GetOwnerReferences(), currentOwner) {
+			continue
+		}
+		// avoid taking ownership of the bootstrap data secret
+		if s.Name == currentOwner.GetName() {
+			continue
+		}
+
+		ss := s.DeepCopy()
+
+		ss.SetOwnerReferences(util.ReplaceOwnerRef(ss.GetOwnerReferences(), metav1.OwnerReference{
+			APIVersion:         controlplanev1.GroupVersion.String(),
+			Kind:               "KubeadmControlPlane",
+			Name:               kcp.Name,
+			UID:                kcp.UID,
+			Controller:         pointer.BoolPtr(true),
+			BlockOwnerDeletion: pointer.BoolPtr(true),
+		}, currentOwner))
+
+		if err := r.Client.Update(ctx, ss); err != nil {
```

Comment:
this should likely be a patch instead of an update, but i think @detiber already mentioned that

Reply:
addressed in a comment that Patch uses a merge strategy that this type doesn't want